File: frontend/src/api/todos.ts
import { api } from "@/api/client";
import type { Todo, TodoPayload } from "@/types";

export const fetchTodos = () => api.get<Todo[]>("/todos");

export const createTodo = (payload: TodoPayload) => api.post<Todo>("/todos", payload);

export const updateTodo = (id: number, payload: Partial<TodoPayload>) =>
  api.put<Todo>(`/todos/${id}`, payload);

export const deleteTodo = (id: number) => api.delete(`/todos/${id}`);

export const reorderTodos = (ids: number[]) =>
  api.put<Todo[]>("/todos/reorder", { ids });

File: backend/app/routers/todos.py
from fastapi import APIRouter

from app.api.deps import DbSession
from app.schemas.todo import TodoCreate, TodoRead, TodoReorder, TodoUpdate
from app.services import todo_service

router = APIRouter(prefix="/todos", tags=["todos"])


@router.get("", response_model=list[TodoRead])
def list_todos(db: DbSession) -> list[TodoRead]:
    return [TodoRead.model_validate(t) for t in todo_service.list_todos(db)]


@router.post("", response_model=TodoRead, status_code=201)
def create_todo(payload: TodoCreate, db: DbSession) -> TodoRead:
    return TodoRead.model_validate(todo_service.create_todo(db, payload))


# Declared before /{todo_id} so "reorder" is not parsed as an id.
@router.put("/reorder", response_model=list[TodoRead])
def reorder_todos(payload: TodoReorder, db: DbSession) -> list[TodoRead]:
    return [
        TodoRead.model_validate(t)
        for t in todo_service.reorder_todos(db, payload.ids)
    ]


@router.put("/{todo_id}", response_model=TodoRead)
def update_todo(todo_id: int, payload: TodoUpdate, db: DbSession) -> TodoRead:
    return TodoRead.model_validate(todo_service.update_todo(db, todo_id, payload))


@router.delete("/{todo_id}", status_code=204)
def delete_todo(todo_id: int, db: DbSession) -> None:
    todo_service.delete_todo(db, todo_id)
